chain/network-primitives/src/types.rs

```diff
@@ -319,6 +321,61 @@ pub enum NetworkViewClientMessages {
     AnnounceAccount(Vec<(AnnounceAccount, Option<EpochId>)>),
 }
 
+#[derive(Debug)]
+pub struct WithHash<T> {
+    inner: T,
+    hash: u64,
+}
+
+impl<T: Hash> Hash for WithHash<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_u64(self.hash);
+    }
+}
+
+impl<T: Hash + Default> Default for WithHash<T> {
+    fn default() -> Self {
+        WithHash::new(T::default())
+    }
+}
+
+impl<T> std::cmp::PartialEq for WithHash<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
```

Comment:
This... feels potentially wrong? hash very much does not guarantee that different values hash to different u64s, especially given that we don't randomize the hash seed. This is also confusing, as I assumed that `WithHash` is just for caching the hash, and almost missed overridden `eq`.

Reply:
Whole blockchain technology relies on hashes not colliding :P. What if I put here a sha256 instead? I'm using a weaker hash here because there is no adversary. Also the number of pairs of values compared is order of magnitude smaller than what is usually required for hashing functions. To make it a truly correct probabilisitic algo, I could select a specific hashing algorithm (rather than whatever hashmap uses) with decent properties and seed it at random on binary startup. Other alternatives that I see are as follows:
1. Go back to identifying sets with EpochId and have an unenforced invariant that EpochId determines the set.
2. Just always do a full comparison and have an unenforced invariant that ClientActor won't send us updates more often than necessary.

Reply:
Then Arc::ptr_eq it is.